src/TestFramework/MSTest.Core/Attributes/DataTestMethodAttribute.cs

```diff
@@ -24,40 +25,36 @@ public class DataTestMethodAttribute : TestMethodAttribute
         /// </returns>
         public override TestResult[] Execute(ITestMethod testMethod)
         {
-            DataRowAttribute[] dataRows = testMethod.GetAttributes<DataRowAttribute>(false);
+            TestDataSource[] dataSources = testMethod.GetAttributes<TestDataSource>(true);
 
-            if (dataRows == null || dataRows.Length == 0)
+            if (dataSources == null || dataSources.Length == 0)
             {
                 return new TestResult[] { new TestResult() { Outcome = UnitTestOutcome.Failed, TestFailureException = new Exception(FrameworkMessages.NoDataRow) } };
             }
 
-            return RunDataDrivenTest(testMethod, dataRows);
+            return RunDataDrivenTest(testMethod, dataSources);
         }
 
         /// <summary>
         /// Run data driven test method.
         /// </summary>
         /// <param name="testMethod"> Test method to execute. </param>
-        /// <param name="dataRows"> Data Row. </param>
+        /// <param name="testDataSources">Test data sources. </param>
         /// <returns> Results of execution. </returns>
-        internal static TestResult[] RunDataDrivenTest(ITestMethod testMethod, DataRowAttribute[] dataRows)
+        internal static TestResult[] RunDataDrivenTest(ITestMethod testMethod, TestDataSource[] testDataSources)
         {
             List<TestResult> results = new List<TestResult>();
 
-            foreach (var dataRow in dataRows)
+            foreach (var testDataSource in testDataSources)
             {
-                TestResult result = testMethod.Invoke(dataRow.Data);
-
-                if (!string.IsNullOrEmpty(dataRow.DisplayName))
-                {
-                    result.DisplayName = dataRow.DisplayName;
-                }
-                else
+                foreach (var data in testDataSource.GetData(testMethod.MethodInfo))
                 {
-                    result.DisplayName = string.Format(CultureInfo.CurrentCulture, FrameworkMessages.DataDrivenResultDisplayName, testMethod.TestMethodName, string.Join(",", dataRow.Data));
-                }
+                    TestResult result = testMethod.Invoke(data);
```

Comment:
Here is how it would look like : 
![image](https://user-images.githubusercontent.com/13763032/26960504-bf42de6c-4c8c-11e7-85ec-f46b2e17d728.png)